fmn/cache/rules.py

```diff
@@ -0,0 +1,52 @@
+import logging
+from typing import TYPE_CHECKING
+
+from cashews import cache
+from cashews.formatter import get_templates_for_func
+
+from ..database.model import Rule
+from .util import cache_early_ttl, cache_ttl
+
+if TYPE_CHECKING:
+    from fedora_messaging.message import Message
+    from sqlalchemy.ext.asyncio import AsyncSession
+
+log = logging.getLogger(__name__)
+
+
+class RulesCache:
+    """Cache the rules currently in the database."""
+
+    def __init__(self):
+        self._db: "AsyncSession" | None = None
+
+    def set_db(self, db: "AsyncSession"):
+        self._db = db
+
+    @cache.locked(key="rules", prefix="v1", ttl="1h")
+    async def build(self):
+        if self._db is None:
+            raise RuntimeError("You must use set_db() first.")
+        log.debug("Building the rules cache")
+        result = await self._db.execute(Rule.select_related().filter_by(disabled=False))
+        log.debug("Built the rules cache")
+        return list(result.scalars())
+
+    @cache.early(
+        key="rules", prefix="v1", ttl=cache_ttl("rules"), early_ttl=cache_early_ttl("rules")
+    )
+    async def get_rules(self):
+        return await self.build()
```

Comment:
Do we use the `build()` method directly? Because we could apply both decorators to the same method (or use `lock=True` on just the `@cache.early()` decorator) according to the [Cashews docs](https://github.com/Krukov/cashews/#locked).

Reply:
Yeah, I might merge those two functions. It was practical for unit tests, but I suppose I can do without. I'll keep a second decorator though, otherwise the lock lifetime does not have a TTL.